pkg/controlplane/server.go

```diff
@@ -126,9 +128,67 @@ func CreateServerChain(completedOptions completedServerRunOptions, stopCh <-chan
 		}
 	})
 
-	return aggregatorServer, nil
+	downloader := aggregator.NewDownloader()
+
+	aggregatorServer.GenericAPIServer.Handler.NonGoRestfulMux.HandleFunc("/openapi/v2", func(res http.ResponseWriter, req *http.Request) {
+		req = req.Clone(req.Context())
+		req.URL.Path = "/openapi/v2"
+		req.URL.RawPath = req.URL.Path
+		
+		cluster := genericapirequest.ClusterFrom(req.Context())
+
+		withCluster := func(handler http.Handler) http.HandlerFunc {
+			return func(res http.ResponseWriter, req *http.Request) {
+				if cluster != nil {
+					req = req.Clone(genericapirequest.WithCluster(req.Context(), *cluster))
+				}
+				handler.ServeHTTP(res, req)
+			}
+		}
+
+		// DAVID TODO: try to understand why adding the cluster returns nil here. The default openAPIService should work 
+		controlPlaneSpec, etag, httpStatus, err := downloader.Download(kubeAPIServer.GenericAPIServer.Handler.Director, "")
+		klog.Infof(etag, httpStatus)
+
+		crdSpecs, etag, httpStatus, err := downloader.Download(withCluster(apiExtensionsServer.GenericAPIServer.Handler.Director), "")
+		klog.Infof(etag, httpStatus)
+
+		mergedSpecs, err := builder.MergeSpecs(controlPlaneSpec, crdSpecs)
+
+		openAPIVersionedService, err := handler.NewOpenAPIService(mergedSpecs)
+		if err != nil {
+			klog.Errorf("Error while building OpenAPI schema: %v", err)
+		}
+
+		handler := &singlePathHandler{}
+
+		err = openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", handler)
```

Comment:
What's the point of this?  Is this registering the service lazily?  Generally I'd avoid dynamic registration and instead have an endpoint that can deal with delegation inline.  Either way, comment on the *why* of registering.

Reply:
I'm trying to reuse `kube-openapi` API as much as possible..